src/bthome_ble/parser.py

```diff
@@ -555,17 +560,30 @@ def _decrypt_bthome(
         else:
             uuid = b"\xd2\xfc" + bytes([adv_info])
         encrypted_payload = data[:-8]
-        count_id = data[-8:-4]
+        last_encryption_counter = self.encryption_counter
+        self.encryption_counter = data[-8:-4]
         mic = data[-4:]
 
-        # nonce: mac [6], uuid16 [2 (v1) or 3 (v2)], count_id [4]
-        nonce = b"".join([bthome_mac, uuid, count_id])
+        # nonce: mac [6], uuid16 [2 (v1) or 3 (v2)], encryption_counter [4]
+        nonce = b"".join([bthome_mac, uuid, self.encryption_counter])
 
         associated_data = None
         if sw_version == 1:
             associated_data = b"\x11"
 
         assert self.cipher is not None  # nosec
+
+        # verify that the encryption counter is increasing compared the previous value
+        if (
+            self.encryption_counter < last_encryption_counter
+            and last_encryption_counter < b"\xF0\xFF\xFF\xFF"
```

Comment:
From a purely security perspective, I'd prefer to get the warning after an overflow. It's a good indication that it is time to re-generate the bind key on the device. That way all the new counter values are encrypted with a different key and replays will not be possible.

Also note that these are little-endian values.

```
>>> hex(int.from_bytes(b'\xF0\xFF\xFF\xFF', "little", signed=False))
'0xfffffff0'
```

Reply:
Changed it. It should now raise a warning if the `encryption_counter` is reaching the maximum `0xFFFFFFFF` = 4294967295 minus 100 = 4294967195. 

For now, after the counter is restarted at `0x00000000`, you can just restart HA/reload the integration to get the sensor working in HA again (as the `encryption_counter` always starts at zero after a restart).